guidance/models/transformers/_transformers_phi3v.py

```diff
@@ -0,0 +1,218 @@
+    Modality
+)
+from guidance.models.transformers._transformers import TransformersTokenizer
+
+try:
+    from PIL import Image
+    has_pillow = True
+except ModuleNotFoundError:
+    has_pillow = False
+
+logger = logging.getLogger(__name__)
+
+
+class TransformersPhi3VisionEngine(Engine):
+    def __init__(
+        self,
+        model="microsoft/Phi-3-vision-128k-instruct",
+        compute_log_probs=False,
+        **kwargs,
+    ):
+        if not has_pillow:
+            raise Exception("Please install pillow with `pip install pillow` to use Phi 3 Vision")
+        self.model_name = model
+        # Initialize the underlying Phi 3 Vision model
+        self.model_obj = AutoModelForCausalLM.from_pretrained(model, **kwargs)
+        self.device = self.model_obj.device
+
+        # Processor handles tokenization and image processing
+        self.processor = AutoProcessor.from_pretrained(self.model_name, trust_remote_code=True)
+        self.tokenizer = TransformersTokenizer(model, self.processor.tokenizer, sp_whitespace=True)
+        super().__init__(self.tokenizer, compute_log_probs)
+
+        # Cache for past key values
+        self._past_key_values = None
+        self._cached_token_ids: list[int] = []
+
+
+    def start(self, prompt, grammar, media: dict, ensure_bos_token=True) -> TokenParser:
+        if isinstance(prompt, bytes):
+            prompt = prompt.decode("utf-8")
+        elif isinstance(prompt, str):
+            prompt = prompt
+        elif isinstance(prompt, TokenParser):
+            raise NotImplementedError(
+                "Still need to implement support for extending a full Parser state."
+            )
+        else:
+            raise Exception("The passed prompt is of an unknown type!")
+
+        # Map Guidance placeholders to Phi 3 Vision format
+        # and make list of images for processing
+        images = []
+        processed_prompt = prompt
+        matches = {}
+        for match in modality_pattern.finditer(prompt):
+            match_str = match.group(0)
+            modality_type = match.group(1)
+            if modality_type != Modality.IMAGE.name:
+                logger.debug("Skipping non-image modality: %s", match_str)
+                continue
+            media_id = match.group(2)
+            if match_str not in matches:
+                matches[match_str] = media_id 
+
+        image_counter = 1
+        for match in matches.keys():
+            processed_prompt = processed_prompt.replace(
+                match, f"<|image_{image_counter}|>"
+            )
+            media_key = matches[match]
+            images.append(Image.open(io.BytesIO(media[media_key])))
+            image_counter += 1
+        logger.debug("Transformed prompt: %s -> ", prompt, processed_prompt)
+
+        model_inputs = self.processor(
+            text=processed_prompt,
+            images=images if len(images) > 0 else None,
+            return_tensors="pt",
+        ).to(self.device)
```

Comment:
Let's make sure that this gets called if images are added again later, after the model starts generating. Otherwise we should trigger a "re-process" upon the next user image being inserted right?

Reply:
Could be good to add some tests using `unittest.patch` if there are any assertions around how many times this should be called for a given image, what should be reused, etc.

Reply:
I believe we'll have to re-process when new images are added. I'll look into how to tell if re-processing is necessary. For the moment, I actually haven't seen a Transformers model yet that was trained to support multiple images during inference. As far as I can tell, Phi 3 vision and llama 3.2 were trained for 1 image per context. That being said, we should still support multi-image scenarios here.